### ISSUE TITLE
Application crashes

### ISSUE THREAD

**Issue by aiolos01** [gerwaric/acquisition#39]
Several bugs to report. First of all I get a dialog saying the latest version is 0.11 despite already using 0.11

Second it crashes every time I choose refresh checked or update all tabs.  If I select some tabs and choose refresh selected it doesn't crash.

Third, I got a different crash once.
![Snap247](https://github.com/gerwaric/acquisition/assets/6642661/f0d88b22-f588-498b-9aac-2d40f8442e4a)

**Comment by aiolos01** [comment]
Hi,
I've run it twice.Once refreshed all checked and once refreshed all tabs.
Both cause immediate crashes. I can't see anything useful in the log. Is
there some other log?

INFO  2024-07-03T01:18:01.246
-------------------------------------------------------------------------------

INFO  2024-07-03T01:18:01.247 acquisition 0.11.0.0 ( version code 61 )
INFO  2024-07-03T01:18:01.247 Built with Qt 6.5.3 on Wed Jun 19 21:10:30
2024
INFO  2024-07-03T01:18:01.247 Running on Qt 6.5.3
INFO  2024-07-03T01:18:01.247 Logging level is INFO
INFO  2024-07-03T01:18:01.315 Running application...
INFO  2024-07-03T01:18:01.383 OAuth: refreshing token at "Mon Jun 24
11:33:57 2024 GMT+0300"
INFO  2024-07-05T18:17:04.748
-------------------------------------------------------------------------------

INFO  2024-07-05T18:17:04.751 acquisition 0.11.1.0 ( version code 61 )
INFO  2024-07-05T18:17:04.751 Built with Qt 6.5.3 on Mon Jul 1 22:12:12
2024
INFO  2024-07-05T18:17:04.751 Running on Qt 6.5.3
INFO  2024-07-05T18:17:04.751 Logging level is INFO
INFO  2024-07-05T18:17:04.753 Running application...
INFO  2024-07-05T18:17:04.764 Removing the stored OAuth token because it
has expired.
WARN  2024-07-05T18:17:28.579 Constructing OAuth token without a birthday.
WARN  2024-07-05T18:17:28.579 Constructing OAuth token without an
expiration
INFO  2024-07-05T18:17:28.594 OAuth: refreshing token at "Sat Jul 6
03:17:28 2024"
INFO  2024-07-05T18:17:36.900 Starting OAuth authentication
INFO  2024-07-05T18:17:37.682 I've created the folder  "....."
INFO  2024-07-05T18:17:37.796 Logging level set to "INFO"
INFO  2024-07-05T18:17:38.176 Loading item classes from RePoE.
INFO  2024-07-05T18:17:38.652 Loading item base types from RePoE.
INFO  2024-07-05T18:17:39.944 There are 6629 items and 203 tabs after the
refresh.
INFO  2024-07-05T18:17:39.944 There are 33  uncategorized items.
INFO  2024-07-05T18:19:51.672
-------------------------------------------------------------------------------

INFO  2024-07-05T18:19:51.683 acquisition 0.11.1.0 ( version code 61 )
INFO  2024-07-05T18:19:51.683 Built with Qt 6.5.3 on Mon Jul 1 22:12:12
2024
INFO  2024-07-05T18:19:51.683 Running on Qt 6.5.3
INFO  2024-07-05T18:19:51.683 Logging level is INFO
INFO  2024-07-05T18:19:51.685 Running application...
INFO  2024-07-05T18:19:51.693 OAuth: refreshing token at "Sat Jul 6
03:17:28 2024 GMT"
INFO  2024-07-05T18:19:53.925 Starting OAuth authentication
INFO  2024-07-05T18:19:54.661 Logging level set to "INFO"
INFO  2024-07-05T18:19:54.842 Loading item classes from RePoE.
INFO  2024-07-05T18:19:55.024 Loading item base types from RePoE.
INFO  2024-07-05T18:19:55.672 There are 6629 items and 203 tabs after the
refresh.
INFO  2024-07-05T18:19:55.672 There are 33  uncategorized items.

**Comment by aiolos01** [comment]
I don't see anything useful in the log with trace on either.

INFO  2024-07-07T19:11:20.807 Logging level set to "TRACE"
DEBUG 2024-07-07T19:11:33.527 Updating Checked stash tabs
DEBUG 2024-07-07T19:11:33.527 Updating 89  tabs.
DEBUG 2024-07-07T19:11:33.527 Keeping 114 tabs and culling 89
DEBUG 2024-07-07T19:11:33.527 Keeping 0 items and culling 6585
DEBUG 2024-07-07T19:11:33.765 Creating rate limit policy
"stash-list-request-limit" for "GET /stash/<league>"
DEBUG 2024-07-07T19:11:33.765 "stash-list-request-limit" increasing history
capacity from 0 to 30
TRACE 2024-07-07T19:11:33.765 RateLimiter is PAUSED 0 for ""
TRACE 2024-07-07T19:11:33.765 "stash-list-request-limit" waiting 0 seconds
to send request 1 at "Sun Jul 7 19:11:33 2024"
TRACE 2024-07-07T19:11:33.765 RateLimiter is PAUSED 0 for
"stash-list-request-limit"
TRACE 2024-07-07T19:11:34.017 "stash-list-request-limit" sending request 1
to "GET /stash/<league>" via "https://api.pathofexile.com/stash/Standard"
TRACE 2024-07-07T19:11:34.259 "stash-list-request-limit" received reply for
request 1 with status 200
TRACE 2024-07-07T19:11:34.259 Updating next send: from "Sun Jul 7 19:11:33
2024" to "Sun Jul 7 19:11:34 2024"
TRACE 2024-07-07T19:11:34.260 RateLimiter is PAUSED 0 for
"stash-list-request-limit"
TRACE 2024-07-07T19:11:34.260 OAuth stash list received
DEBUG 2024-07-07T19:11:34.260 Received stash list, there are 161 stash tabs
DEBUG 2024-07-07T19:11:34.260 Queued ( 1 ) --  "#1, \"...\""
DEBUG 2024-07-07T19:11:34.260 Queued ( 2 ) --  "#2, \" ... \""
DEBUG 2024-07-07T19:11:34.260 Queued ( 3 ) --  "#3, \" ... \""
DEBUG 2024-07-07T19:11:34.260 Queued ( 4 ) --  "#4, \" ... \""
DEBUG 2024-07-07T19:11:34.260 Queued ( 5 ) --  "#5, \" ... \""
DEBUG 2024-07-07T19:11:34.260 Queued ( 6 ) --  "#6, \" ... \""
DEBUG 2024-07-07T19:11:34.260 Queued ( 7 ) --  "#7, \" ... \""
DEBUG 2024-07-07T19:11:34.260 Queued ( 8 ) --  "#8, \" ... \""

It seems to crash after queuing the currency tab. I don't know if that
helps in some way. I have an account on discord, not that I ever use it.
I'll try to PM you there.

**Comment by aiolos01** [comment]
OK I've run it. This one crashes on launch. I assume crashpad is the
utility that sends the report, I hope it helps.

**Comment by aiolos01** [comment]
I believe I've sent a couple of reports with the new version

**Comment by gerwaric** [comment]
Yes. I can see a crash with alpha-trace.3 that looks plausible. This may have sold me on BugSplat, because it would have taken a long time to find the issue without it.

Specifically, the crash occured while trying to parse the colour of a stash tab after receiving the list of stashes in OAuth mode. That field is optional according to the GGG API developer documentation. My OAuth code assumed it was always present. This wasn't happening to me, so apparently all my tabs have colour.

If you were experiencing this in Standard, it's possible that stash tabs created in the early days of PoE don't have that colour field. This problem is also a non-issue when using POESESSID because the legacy API returns stash tabs in a different json format.

Long story short--this should fix that particular crash:
https://github.com/gerwaric/acquisition/releases/tag/v0.11.1-colour-fix

While you're testing that, I've been down in the guts of the items model trying to diagnose the "bucket row out of bounds error." It's making my head hurt, but that's the next thing I'm working on.

**Comment by aiolos01** [comment]
Yes I'm on standard. Unfortunately this still crashes in both SSID and OATH
login options. Sent 2 reports.

**Comment by gerwaric** [comment]
One of those crashes is showing up on BugSplat, and it looks the crash was outside acquisition in the Microsoft Visual C++ runtime library `msvcp140.dll`:
```
d:\a01\_work\12\s\src\vctools\crt\github\stl\src\mutex.cpp(103):
mtx_do_lock(_Mtx_internal_imp_t*, xtime const*)
```

Can you check to see which version(s) of the c++ runtime you have installed?

Here's what's on my development system:
![image](https://github.com/gerwaric/acquisition/assets/6461624/5304fcf6-015f-4a79-a5c9-dd0ac20826b4)

**Comment by aiolos01** [comment]
The latest installed is 14.36.33130.
Inside the acquisition install folder are:
msvcp140.dll   14.29.30139.0
msvcp140_1.dll   14.29.30139.0
msvcp140_2.dll   14.29.30139.0

**Comment by gerwaric** [comment]
I'll see if I can roll back to those versions and replicate the bugs you're seeing. If that works, then I can link to the latest installer from Microsoft and mention it in the release notes. I'll update you here.

**Comment by gerwaric** [comment]
Please delete those files from the installation folder (`msvcp140.dll`, `msvcp140_1.dll`, and `msvcp140_2.dll`) and let me know what happens.

I installed multiple versions of the visual c++ runtime one at a time, from 14.20.27519 to 14.40.33810, including 14.29.30139. Each time I tried running acquisition both with and without those three files from 14.29.30139, and acquisition crashed every time those .dlls were present in the installation folder.

**Comment by aiolos01** [comment]
I deleted them and now I can run acquisition. With SSID I can update the
tabs. With OATH it crashes when I try to update (tried update all). Report
sent.

**Comment by gerwaric** [comment]
Whew. I'm glad that SSID works for you again. I'll see what I can figure out from those crash reports and try to get another alpha release soon.

**Comment by gerwaric** [comment]
The next release has extra checks when parsing the OAuth json, so if that was causing the crashes you are seeing, they will at least show up in the log file as warnings or errors.

https://github.com/gerwaric/acquisition/releases/tag/v0.11.1-alpha.6

This update should also warn you if there are extra dll files in the application folder, since there may be other users experiencing this crash.

**Comment by aiolos01** [comment]
Ok this seems to work when updating tabs with OATH. However it crashes in
searches with OATH. The same search works with SSID. I find this weird
since searching is done locally but I guess it's not as simple as that.

**Comment by aiolos01** [comment]
Scratch that. It crashes with SSID too during search.

**Comment by gerwaric** [comment]
> Scratch that. It crashes with SSID too during search.

Oof. I'm sorry. It's possible that once you did a refresh with OAuth the data that caused the crash was saved into the data files. This could be the reason you didn't see a crash with SSID at first.

Fortunately I can see those crashes in bugsplat.

When I get up in a couple hours (I'm on US mountain time), I'll dig into those reports. Hopefully they will help identify whatever elusive thing is happening.

Another thought I have is that if it's crashing in offline search for you, then I should be able to replicate the crash if I had your json data. I don't like the idea of asking you to share data files directly, so I might add an "export" function if the crash reports aren't enough.

Can't stress enough how grateful I am at your persistence, since I can't replicate these crashes on my own. Thank you very much.

**Comment by aiolos01** [comment]
Just realised the crashes might have been my fault. Acquisition installed a
new VC redist version but I didn't reboot because I was busy with other
things. I rebooted now and after some light testing it didn't crash. I'll
test it some more and get back to you.
Thanks for continuing to work on this.

**Comment by gerwaric** [comment]
I'm glad to hear the reboot helped. From the crash reports, I think we might be back to your "bucket row out of bounds" error. If you can reliably reproduce that error, a short writeup of how would be useful. I've been playing around with searches, but think I've only seen that error once, so I need to do some more poking around.

**Comment by gerwaric** [comment]
@aiolos01 I think I may have solved the "bucket row out of bounds error":

https://github.com/gerwaric/acquisition/releases/tag/v0.11.1-alpha.8

The issue I found is that acquisition always assumed that when the selected item changed, it was changing to a new item. However, it is possible that the change meant that no item was selected. When that happens, an internal row index is negative, which caused the error you screenshotted at the start of this thread. Now acquisition checks for that case and should gracefully do nothing.

When you can get around to testing this, let me know if it's working, or if there are any other new issues for you.

**Comment by gerwaric** [comment]
@aiolos01 I found a new place where the same bucket row error happens. If I missed that one, I likely missed several, so I'm going to go through the code with a fine-toothed comb tomorrow.  I'll reply here when I have an update.

**Comment by gerwaric** [comment]
I have a lot to say about this issue, but I think it might finally be fixed:

https://github.com/gerwaric/acquisition/releases/tag/v0.11.1-alpha.10

First, I was able to replicate a search crash by clicking on the "corrupted" checkbox, then selected a search result that was not priced, then clicking on the "priced" checkbox, which should cause the item I had selected to be removed from the search results. It turns out I was incorrectly clearing the selection object after the search was updated, so it was making calls with invalid arguments. This is fixed.

Second, I'm not sure that is the only source of those mysterious search crashes. However, the common issue seems to be certain function calls that lead to dereference an invalid pointer, which is a big no-no. To forestall any future crashes, I added extra code to make sure those functions aren't called with invalid arguments. Now, acquisition will log and error and continue running.

Finally, I think the guts of the item searching needs to be refactored. In the ten years since acquisition was created, people now have millions of items, which acquisition was not designed to manager. There are new frameworks that may be simpler, faster, and fewer lines of code. This would be a lot of work, so it's on the back burner for now.

**Comment by aiolos01** [comment]
I wouldn't worry about speed. I have lots of tabs (ok not as many as people
who play every league) and it's very fast. If the old framework has memory
issues (or something else that breaks the app) then it should be fixed,
otherwise I wouldn't consider it a priority.

**Comment by gerwaric** [comment]
That's good feedback. Since acquisition isn't popular these days, the only feedback I get about what features people use or care about comes from the ones willing to use GitHub to post issues.

Part of me wants to rebuild a new acquisition from scratch given changes GGG has made, and changes in technology over the last decade. The current guts are pretty complicated and tricky to understand. However, a more practical priority is to improve the mod searching, since it doesn't distinguish between implicit, explicit, and other modifier types the way that the trade site does. It also doesn't compute pseudo-mods at all. (It looks like it might have worked at some point in the past, but right now it's not).

**Comment by aiolos01** [comment]
To be honest I don't use acquisition to list items any more. The premium
tabs are so easy to use. You don't have to type the name of the item when
you get the PM, the tab is listed in the message and the item is even
highlighted in the tab. There no way to beat that. It's great that
acquisition can read the prices though. It's still useful to see them all
in one place without having to hover over each item. The only items I still
have listed through acquisition are from ancient leagues that I can't be
bothered to relist.

However acquisition is still invaluable as a personal search tool. I often
play in standard and I have a ton of stuff. There is no way I could find
anything without acquisition (or looty). Even in leagues it's easier to
just look for things there. So I believe this is the direction that you
should take the app. Make it a better search tool. I don't expect it to
rival the search engine but there's definite room for improvement there if
you have time.Pseudo-mods are a great starting point.

**Comment by gerwaric** [comment]
That's great to hear.  Don't tell anyone just yet, but my long-term vision is to have three types of searches:

1. Equipable items and jewels
2. Stackable items, aka currency
3. Unstackables, such as maps and logbooks

However, it's probalby more immediately useful to get pseudo mods working. There's code that suggests they worked some time in the past, but I haven't looked into when that stopped working. Heck, it's even possible some of the early changes I made to acquisition did it, because I was learning c++ as well as acquisition's code base and GGG's API and data structures.

Coming back to the topic of this thread...

If you don't have any issues with the latest alpha, I'd like to release v0.11.1 in the next week or two. (Caveat: https://github.com/gerwaric/acquisition/issues/50 may become a blocking issue).

**Comment by aiolos01** [comment]
No issues so far. I'll let you know if I find any.

**Comment by gerwaric** [comment]
A few hours ago another user had their API access blocked at the network level due to a previously unknown bug in acquisition (https://github.com/gerwaric/acquisition/issues/50). Details along with an update here:

https://github.com/gerwaric/acquisition/releases/tag/v0.11.1-alpha.12

I strongly recommend installing this update to avoid losing access to third-party tools until GGG has time to repsond to any support requests you'd have to send.

**Comment by aiolos01** [comment]
Yeah, I saw that. Damn that 0.11 version is cursed, I suggest skipping it
altogether.

**Comment by gerwaric** [comment]
I have an idea of what happened now. If this build fixes it I'll write it up in more detail:

https://github.com/gerwaric/acquisition/releases/tag/v0.11.1.1-alpha.1

I'm also closing this issue for now, since I think the issues causing the crashes you were reporting have been fixed:
- "bucket row out of bounds"
- incompatible MSVC libraries

If these issues come back, please reopen this issue.
